metadata/src/main/java/org/apache/kafka/controller/metrics/ControllerServerMetrics.java

```diff
@@ -0,0 +1,211 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller.metrics;
+
+import com.yammer.metrics.core.Gauge;
+import com.yammer.metrics.core.MetricName;
+import com.yammer.metrics.core.MetricsRegistry;
+import org.apache.kafka.server.metrics.KafkaYammerMetrics;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * These are the metrics which are managed by the ControllerServer class. They generally pertain to
+ * aspects of the metadata, like how many topics or partitions we have.
+ * All of these except MetadataErrorCount are managed by ControllerServerMetricsPublisher.
+ *
+ * IMPORTANT: Metrics which are managed by the QuorumController class itself should go in
+ * @link{org.apache.kafka.controller.metrics.QuorumControllerMetrics}, not here.
+ */
+public final class ControllerServerMetrics implements AutoCloseable {
+    private final static MetricName FENCED_BROKER_COUNT = getMetricName(
+        "KafkaController", "FencedBrokerCount");
+    private final static MetricName ACTIVE_BROKER_COUNT = getMetricName(
+        "KafkaController", "ActiveBrokerCount");
+    private final static MetricName GLOBAL_TOPIC_COUNT = getMetricName(
+        "KafkaController", "GlobalTopicCount");
+    private final static MetricName GLOBAL_PARTITION_COUNT = getMetricName(
+        "KafkaController", "GlobalPartitionCount");
+    private final static MetricName OFFLINE_PARTITION_COUNT = getMetricName(
+        "KafkaController", "OfflinePartitionsCount");
+    private final static MetricName PREFERRED_REPLICA_IMBALANCE_COUNT = getMetricName(
+        "KafkaController", "PreferredReplicaImbalanceCount");
+    private final static MetricName METADATA_ERROR_COUNT = getMetricName(
+        "KafkaController", "MetadataErrorCount");
+
+    private final Optional<MetricsRegistry> registry;
+    private final AtomicInteger fencedBrokerCount = new AtomicInteger(0);
+    private final AtomicInteger activeBrokerCount = new AtomicInteger(0);
+    private final AtomicInteger globalTopicCount = new AtomicInteger(0);
+    private final AtomicInteger globalPartitionCount = new AtomicInteger();
+    private final AtomicInteger offlinePartitionCount = new AtomicInteger();
+    private final AtomicInteger preferredReplicaImbalanceCount = new AtomicInteger();
+    private final AtomicInteger metadataErrorCount = new AtomicInteger(0);
+
+    /**
+     * Create a new ControllerServerMetrics object.
+     *
+     * @param registry  The metrics registry, or Optional.empty if this is a test and we don't have one.
+     */
+    public ControllerServerMetrics(Optional<MetricsRegistry> registry) {
```

Comment:
I was sort of confused by this, but I think I get it. Since MetricsRegistry is not an interface, we can't easily stub it out for tests. So it's either something like this, or using a mocking library.

Reply:
Yeah. Looking at the previous code, I realized that `MockControllerMetrics` was implementing all the same stuff as `QuorumControllerMetrics` (atomic adds, atomic gets, etc.) So there was no point... just don't register with the registry if you don't care about the registry, and you can avoid the whole layer of abstraction.